In [1]:
#from google.colab import drive
#drive.mount('/content/drive')


In [2]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import create_engine

In [3]:
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
             'pwd': 'Sdf4$2;d-d30pp', # пароль
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
             'port': 6432, # порт подключения
             'db': 'data-analyst-final-project-db'} # название базы данных

In [4]:
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                         db_config['pwd'],
                                                         db_config['host'],
                                                         db_config['port'],
                                                         db_config['db'])

In [5]:
pip install psycopg2-binary

Note: you may need to restart the kernel to use updated packages.


In [6]:
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

## SQL

Коронавирус застал мир врасплох, изменив привычный порядок вещей. В
свободное время жители городов больше не выходят на улицу, не посещают кафе
и торговые центры. Зато стало больше времени для книг. Это заметили
стартаперы — и бросились создавать приложения для тех, кто любит читать.

Ваша компания решила быть на волне и купила крупный сервис для чтения книг
по подписке. Ваша первая задача как аналитика — проанализировать базу
данных.

В ней — информация о книгах, издательствах, авторах, а также пользовательские
обзоры книг. Эти данные помогут сформулировать ценностное предложение для
нового продукта.

### Таблица books
Содержит данные о книгах:
- book_id — идентификатор книги;
- author_id — идентификатор автора;
- title — название книги;
- num_pages — количество страниц;
- publication_date — дата публикации книги;
- publisher_id — идентификатор издателя.

### Таблица authors
Содержит данные об авторах:
- author_id — идентификатор автора;
- author — имя автора.

### Таблица publishers
Содержит данные об издательствах:
- publisher_id — идентификатор издательства;
- publisher — название издательства;

### Таблица ratings
Содержит данные о пользовательских оценках книг:
- rating_id — идентификатор оценки;
- book_id — идентификатор книги;
- username — имя пользователя, оставившего оценку;
- rating — оценка книги.

### Таблица reviews
Содержит данные о пользовательских обзорах на книги:
- review_id — идентификатор обзора;
- book_id — идентификатор книги;
- username — имя пользователя, написавшего обзор;
- text — текст обзора.

## Задания
- Посчитайте, сколько книг вышло после 1 января 2000 года;
- Для каждой книги посчитайте количество обзоров и среднюю оценку;
- Определите издательство, которое выпустило наибольшее число книг толще
50 страниц — так вы исключите из анализа брошюры;
- Определите автора с самой высокой средней оценкой книг — учитывайте
только книги с 50 и более оценками;
- Посчитайте среднее количество обзоров от пользователей, которые поставили
больше 50 оценок.

### Описать цели исследования:

- Основная цель исследования - сформулировать ценностное предложение для нового продукта, сервиса для чтения книг по подписке.
- Для этого необходимо выполнить задания.

## Исследование таблиц.

### Таблица books

In [7]:
query = '''
           SELECT * 
           FROM books
           
        '''
pd.io.sql.read_sql(query, con = engine)

book_id  author_id                                              title  \
0          1        546                                       'Salem's Lot   
1          2        465                 1 000 Places to See Before You Die   
2          3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3          4         82  1491: New Revelations of the Americas Before C...   
4          5        125                                               1776   
..       ...        ...                                                ...   
995      996        571           Wyrd Sisters (Discworld  #6; Witches #2)   
996      997        454                        Xenocide (Ender's Saga  #3)   
997      998        201                                    Year of Wonders   
998      999         94                        You Suck (A Love Story  #2)   
999     1000        509  Zen and the Art of Motorcycle Maintenance: An ...   

     num_pages publication_date  publisher_id  
0          594       2005-11-01            93  
1          992       2003-05-22           336  
2          322       2010-12-21           135  
3          541       2006-10-10           309  
4          386       2006-07-04           268  
..         ...              ...           ...  
995        265       2001-02-06           147  
996        592       1996-07-15           297  
997        358       2002-04-30           212  
998        328       2007-01-16           331  
999        540       2006-04-25           143  

[1000 rows x 6 columns]

В БД есть информация о 1000 книг.

### Таблица authors

In [8]:
query = '''
           SELECT * 
           FROM authors
           
        '''
pd.io.sql.read_sql(query, con = engine)

author_id                                author
0            1                            A.S. Byatt
1            2        Aesop/Laura Harris/Laura Gibbs
2            3                       Agatha Christie
3            4                         Alan Brennert
4            5              Alan Moore/David   Lloyd
..         ...                                   ...
631        632         William Strunk Jr./E.B. White
632        633                           Zadie Smith
633        634                 Zilpha Keatley Snyder
634        635                    Zora Neale Hurston
635        636  Åsne Seierstad/Ingrid Christopherson

[636 rows x 2 columns]

В БД есть информация о 636 авторах.

### Таблица publishers

In [9]:
query = '''
           SELECT * 
           FROM publishers
           
        '''
pd.io.sql.read_sql(query, con = engine)

publisher_id                          publisher
0               1                                Ace
1               2                           Ace Book
2               3                          Ace Books
3               4                      Ace Hardcover
4               5  Addison Wesley Publishing Company
..            ...                                ...
335           336         Workman Publishing Company
336           337                         Wyatt Book
337           338              Yale University Press
338           339                           Yearling
339           340                     Yearling Books

[340 rows x 2 columns]

В БД есть информация о 340 издателях.

### Таблица ratings

In [10]:
query = '''
           SELECT * 
           FROM ratings
           
        '''
pd.io.sql.read_sql(query, con = engine)

rating_id  book_id        username  rating
0             1        1      ryanfranco       4
1             2        1   grantpatricia       2
2             3        1    brandtandrea       5
3             4        2        lorichen       3
4             5        2     mariokeller       2
...         ...      ...             ...     ...
6451       6452     1000  carolrodriguez       4
6452       6453     1000         wendy18       4
6453       6454     1000      jarvispaul       5
6454       6455     1000           zross       2
6455       6456     1000         fharris       5

[6456 rows x 4 columns]

В БД есть информация о 6456 проставленных оценках.

### Таблица reviews

In [11]:
query = '''
           SELECT * 
           FROM reviews
           
        '''
pd.io.sql.read_sql(query, con = engine)

review_id  book_id        username  \
0             1        1    brandtandrea   
1             2        1      ryanfranco   
2             3        2        lorichen   
3             4        3   johnsonamanda   
4             5        3     scotttamara   
...         ...      ...             ...   
2788       2789      999      martinadam   
2789       2790     1000         wknight   
2790       2791     1000  carolrodriguez   
2791       2792     1000         wendy18   
2792       2793     1000      jarvispaul   

                                                   text  
0     Mention society tell send professor analysis. ...  
1     Foot glass pretty audience hit themselves. Amo...  
2     Listen treat keep worry. Miss husband tax but ...  
3     Finally month interesting blue could nature cu...  
4     Nation purpose heavy give wait song will. List...  
...                                                 ...  
2788  Later hospital turn easy community. Fact same ...  
2789   Change lose answer close pressure. Spend so now.  
2790  Authority go who television entire hair guy po...  
2791  Or western offer wonder ask. More hear phone f...  
2792  Republican staff bit eat material measure plan...  

[2793 rows x 4 columns]

В БД есть информация о 2793 отзывах.

### Вывод по таблицам

Всего есть 5 таблиц с данными.

Основная таблица books, с ней связаны другие таблицы посредством первичных и внешних ключей.

В БД содержится информация о:
- 1000 книг
- 636 авторах
- 340 издельствах
- 6456 оценок
- 2793 отзывов

Столбцы таблиц описаны в "описании данных", мы не будем повторяться.

## Выполнение заданий

### Посчитайте, сколько книг вышло после 1 января 2000 года;

In [12]:
query = '''
           SELECT COUNT(title) 
           FROM books
           WHERE CAST(publication_date AS timestamp) > '2000-01-01';
        '''
pd.io.sql.read_sql(query, con = engine)

count
0    819

После 1 января 2000 года вышло 819 книг (но если смотреть уникальные названия title, то 818, возможно у одной книги переиздание?)

### Для каждой книги посчитайте количество обзоров и среднюю оценку;

In [13]:
query = '''
            SELECT b.title AS "Наименование книги", 
                   rv."Количество обзоров",
                   AVG(rt.rating) AS "Средняя оценка"
            FROM books b
            LEFT JOIN (SELECT book_id, 
                              COUNT(text) AS "Количество обзоров"
                       FROM reviews
                       GROUP BY book_id) AS rv ON b.book_id=rv.book_id
            LEFT JOIN ratings rt ON b.book_id=rt.book_id
            --WHERE "Количество обзоров" IS NOT NULL
            GROUP BY b.title, rv."Количество обзоров"
            ORDER BY rv."Количество обзоров" DESC
            --LIMIT 50
        '''

pd.io.sql.read_sql(query, con = engine)

Наименование книги  Количество обзоров  \
0      Anne Rice's The Vampire Lestat: A Graphic Novel                 NaN   
1    Disney's Beauty and the Beast (A Little Golden...                 NaN   
2                            Essential Tales and Poems                 NaN   
3                                 Leonardo's Notebooks                 NaN   
4     The Cat in the Hat and Other Dr. Seuss Favorites                 NaN   
..                                                 ...                 ...   
995                                The Social Contract                 1.0   
996  To Green Angel Tower (Memory  Sorrow  and Thor...                 1.0   
997               Tsubasa: RESERVoir CHRoNiCLE  Vol. 1                 1.0   
998                              Wicked: The Grimmerie                 1.0   
999                  Winter Prey (Lucas Davenport  #5)                 1.0   

     Средняя оценка  
0          3.666667  
1          4.000000  
2          4.000000  
3          4.000000  
4          5.000000  
..              ...  
995        3.500000  
996        4.500000  
997        2.500000  
998        3.500000  
999        4.500000  

[1000 rows x 3 columns]

Больше всего обзоров на книгу Twilight (Twilight #1).

### Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;

In [15]:
query = '''
           SELECT p.publisher AS "Издательство", COUNT(b.book_id) AS "Кол-во книг"
           FROM books b
           LEFT JOIN publishers p ON b.publisher_id=p.publisher_id
           WHERE b.num_pages > 50
           GROUP BY p.publisher
           ORDER BY "Кол-во книг" DESC
           LIMIT 1
        '''
pd.io.sql.read_sql(query, con = engine)

Издательство  Кол-во книг
0  Penguin Books           42

Издательство, которое выпустило наибольшее число книг толще 50 страниц — Penguin Books.

### Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;

In [16]:
query = '''
           SELECT best_rating.author
           FROM (
                 SELECT a.author, COUNT(r.rating), AVG(r.rating)
                 FROM books b
                 JOIN ratings r ON b.book_id=r.book_id
                 JOIN authors a ON b.author_id=a.author_id
                 GROUP BY a.author
                 HAVING COUNT(r.rating) > 50
                 ORDER BY avg DESC) best_rating
           LIMIT 1
        '''
pd.io.sql.read_sql(query, con = engine)

author
0  J.K. Rowling/Mary GrandPré

Автор с самой выской средней оценкой книг - J.K. Rowling/Mary GrandPré.

### Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

In [17]:
query = '''
           SELECT COUNT(rv.text)/COUNT(DISTINCT rv.username) AS avg_review
           FROM reviews rv
           WHERE rv.username IN (SELECT rt.username 
                                 FROM ratings rt
                                 GROUP BY rt.username
                                 HAVING COUNT(rt.rating) > 50 
                                 ORDER BY COUNT(rt.rating) DESC)
           --GROUP BY rv.username

        '''
pd.io.sql.read_sql(query, con = engine)

avg_review
0          24

Среднее количество обзоров от пользователей, поставивших более 50 оценок - **24**.

## Вывод

В базе данных содержится информация о:
- 1000 книг
- 636 авторах
- 340 издельствах
- 6456 оценок книг
- 2793 отзывов о книгах


- После 1 января 2000 года вышло 819 книг (т.е.82% от общего количества в БД).
- Больше всего обзоров на книгу Twilight (Twilight #1).
- Издательство, которое выпустило наибольшее число книг толще 50 страниц — Penguin Books.
- Автор с самой выской средней оценкой книг - J.K. Rowling/Mary GrandPré.
- Среднее количество обзоров от пользователей, поставивших более 50 оценок - 24.

Нужно сформировать ценностное предложение для сервиса. Честно говоря, на основании выполненных заданий ничего в голову стоящего не приходит.

Возможно стоит увеличить базу книг, а то их всего 1000, для "крупного" сервиса на мой взгляд малова-то. При этом на эту 1000 книг приходится 636 авторов, тоже знаете ли малова-то. Нужно привлечь побольше авторов. Что касается пользователей, то они достаточно активные, много оценок и отзывов оставлено. Нужно попытаться не потерять аудиторию, для этого понять что их побуждает оставлять оценки и комменты, мне кажется это редкость. Кстати, 82% книг вышло после 2000 года, я полагаю что имеется ввиду напечатано в конкретном издательстве. Если нет, то возможно стоит добавить классических произведений.